crates/telio-starcast/src/starcast_peer.rs

```diff
@@ -0,0 +1,398 @@
+    const NAME: &'static str = "Starcast";
+
+    /// Error that may occur in [Task]
+    type Err = ();
+
+    /// Wait loop for asynchronously handling events for the component.
+    ///
+    /// # Returns
+    ///
+    /// A response for the runtime indicating when to run again.
+    async fn wait(&mut self) -> WaitResponse<'_, Self::Err> {
+        match self {
+            State {
+                channel: Chan { rx, tx },
+                tunnel: Some(peer),
+                wg_sock_addr: Some(ref wg_sock_addr),
+                socket,
+                ..
+            } => {
+                tokio::select! {
+                    // Periodic timer updates for managing handshakes.
+                    _ = self.wg_timer_tick_interval.tick() => {
+                        match peer.update_timers(&mut self.sending_buffer) {
+                            TunnResult::Done => (),
+                            TunnResult::WriteToNetwork(packet) => {
+                                if let Err(e) = socket.send_to(packet, wg_sock_addr).await {
+                                    telio_log_error!("[Starcast] Update Tunn timers failed to send to socket: {:?}", e);
+                                }
+                            },
+                            TunnResult::Err(e) => {
+                                telio_log_error!("[Starcast] Failed to update Tunn timers: {:?}", e);
+                            }
+                            _ => telio_log_warn!("[Starcast] Update Tunn timers returned unexpected result"),
+                        }
+                    },
+                    // Outbound packets
+                    Some(ip_packet) = rx.recv() => {
+                        match peer.encapsulate(&ip_packet, &mut self.sending_buffer) {
+                            TunnResult::Done => (),
+                            TunnResult::Err(e) => {
+                                telio_log_error!("[Starcast] Failed to send encapsulate packet: {:?}", e);
+                            },
+                            TunnResult::WriteToNetwork(packet) => {
+                                if let Err(e) = socket.send_to(packet, wg_sock_addr).await {
+                                    telio_log_error!("[Starcast] Failed to send encapsulated packet to socket: {:?}", e);
+                                }
+                            },
+                            _ => {
+                                telio_log_warn!("[Starcast] Encapsulate packet returned unexpected result");
+                            },
+                        }
+                    },
+                    // Here we're handling the deferred decapsulation of handshake packets to avoid deadlocking.
+                    _ = socket.writable(), if self.packets_present_in_tunnel => {
+                        match peer.decapsulate(None, &[], &mut self.sending_buffer) {
+                            // Note that the handshake packet here is a slice pointing to the sending buffer.
```

Comment:
Is it possible that the `sending_buffer` contains more than one handshake packet?